pipeline_steps/common.groovy

```diff
@@ -656,28 +656,201 @@ def delete_workspace() {
     usernamePassword(
       credentialsId: "jira_user_pass",
       usernameVariable: "JIRA_USER",
       passwordVariable: "JIRA_PASS"
     )
   ]){
-    sh """#!/bin/bash -xe
+    return sh(script: """#!/bin/bash -xe
+      cd ${env.WORKSPACE}
+      set +x; . .venv/bin/activate; set -x
+      python rpc-gating/scripts/jirautils.py \
+        --user '$JIRA_USER' \
+        --password '$JIRA_PASS' \
+        build_failure_issue \
+          --project "${project}" \
+          --job-name "${env.JOB_NAME}" \
+          --job-url "${env.JOB_URL}" \
+          --build-tag "${env.BUILD_TAG}" \
+          --build-url "${env.BUILD_URL}"
+    """, returnStdout: true)
+  }
+}
+
+
+// Create String of jirautils.py/create_jira_issue options for labels
+// eg: --label "jenkins" --label "maas"
+@NonCPS
+String generate_label_options(List labels){
+  List terms = []
+  for (label in labels) {
+      terms += "--label \"${label}\""
+  }
+  return terms.join(" ")
+}
+
+
+// This method creates a jira issue. Doesn't check for duplicates or
+// do anthing fancy. Use build_failure_issue for build failures.
+String create_jira_issue(String project,
+                      String summary,
+                      String description,
+                      List labels = []){
+  withCredentials([
+    usernamePassword(
+      credentialsId: "jira_user_pass",
+      usernameVariable: "JIRA_USER",
+      passwordVariable: "JIRA_PASS"
+    )
+  ]){
+    return sh (script:"""#!/bin/bash -xe
       cd ${env.WORKSPACE}
       set +x; . .venv/bin/activate; set -x
       python rpc-gating/scripts/jirautils.py \
         --user '$JIRA_USER' \
         --password '$JIRA_PASS' \
         create_issue \
-          --summary "JBF: ${tag}" \
-          --description "Jenkins Build Failed :( [${tag}|${link}]" \
-          --project '$project' \
-          --type '$type'
+          --summary "${summary}" \
+          --description "${description}" \
+          --project "${project}" ${generate_label_options(labels)}
+    """, returnStdout: true)
+  }
+}
+
+
+String get_or_create_jira_issue(String project,
+                              String status = "BACKLOG",
+                              String summary,
+                              String description,
+                              List labels = []){
+  withCredentials([
+    usernamePassword(
+      credentialsId: "jira_user_pass",
+      usernameVariable: "JIRA_USER",
+      passwordVariable: "JIRA_PASS"
+    )
+  ]){
+    return sh (script: """#!/bin/bash -xe
+      cd ${env.WORKSPACE}
+      set +x; . .venv/bin/activate; set -x
+      python rpc-gating/scripts/jirautils.py \
+        --user '$JIRA_USER' \
+        --password '$JIRA_PASS' \
+        get_or_create_issue \
+          --status "${status}" \
+          --summary "${summary}" \
+          --description "${description}" \
+          --project "${project}" ${generate_label_options(labels)}
+    """, returnStdout: true)
+  }
+}
+
+
+List jira_query(String query){
```

Comment:
Where is `jira_query`, `jira_comments`, `jira_close`, `jira_close_all`, etc. used?  I only see them being referenced in the unit tests.

Reply:
They are only used in the unit tests. 

However its difficult to have verify the results of the other functions, without introducing those :(